bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java

```diff
@@ -0,0 +1,206 @@
+
+import javax.annotation.Nullable;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.RecordComponent;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.bson.assertions.Assertions.notNull;
+
+final class RecordCodec<T extends Record> implements Codec<T> {
+    private final Class<T> clazz;
+    private final Constructor<?> canonicalConstructor;
+    private final List<ComponentModel> componentModels;
+    private final ComponentModel componentModelForId;
+    private final Map<String, ComponentModel> fieldNameToComponentModel;
+
+    private record ComponentModel(RecordComponent component, Codec<?> codec, int index) {
+        String getComponentName() {
+            return component.getName();
+        }
+
+        String getFieldName() {
+            if (component.isAnnotationPresent(BsonId.class)) {
+                return "_id";
+            } else if (component.isAnnotationPresent(BsonProperty.class)) {
+                return component.getAnnotation(BsonProperty.class).value();
+            }
+            return getComponentName();
+        }
+
+        Object getValue(final Record record) throws InvocationTargetException, IllegalAccessException {
+            return component.getAccessor().invoke(record);
+        }
+    }
+
+    RecordCodec(final Class<T> clazz, final CodecRegistry codecRegistry) {
+        this.clazz = notNull("class", clazz);
+        canonicalConstructor = notNull("canonicalConstructor", getCanonicalConstructor(clazz));
+        componentModels = getComponentModels(clazz, codecRegistry);
+        fieldNameToComponentModel = componentModels.stream()
+                .collect(Collectors.toMap(ComponentModel::getFieldName, Function.identity()));
+        componentModelForId = getComponentModelForId(clazz, componentModels);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T decode(final BsonReader reader, final DecoderContext decoderContext) {
+        reader.readStartDocument();
+
+        Object[] constructorArguments = new Object[componentModels.size()];
+        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+            var fieldName = reader.readName();
+            var componentModel = fieldNameToComponentModel.get(fieldName);
+            constructorArguments[componentModel.index] = decoderContext.decodeWithChildContext(componentModel.codec, reader);
+        }
+        reader.readEndDocument();
+
+        try {
+            return (T) canonicalConstructor.newInstance(constructorArguments);
+        } catch (ReflectiveOperationException e) {
+            throw new CodecConfigurationException(format("Unable to invoke canonical constructor of record class %s", clazz.getName()), e);
```

Comment:
Seems like the wrong exception, but not sure what the right one is.

Reply:
`CodecConfigurationException` is the grab bag for all manner of codec exceptions